Add initial content for the packaging module

vasiuadriana opened the pull request:
Issue: https://github.com/Kalschatzi/learning_platform/issues/12

vasiuadriana commented on content/module3/3-package-java-app.md:
```diff
@@ -0,0 +1,28 @@
+The standard file format used to package java code is [JAR](https://en.wikipedia.org/wiki/JAR_(file_format)) which stands for Java ARchive.
+In this module, we are going to use the `gradle` tool to create a java executable jar that contains all the dependencies.
+
+<TODO - add here link to the reference app with a basic build.gradle>
```
Hey @tmcalves , I think for this section it would be good to add a reference app, so that we can refer to it, and they can try the packaging on it. 
We can then decide if we use `shadowJar` for example. 
For now, for the next sections, I will just assume that we have a jar already, so that I can carry on filling out the docker content.

tmcalves replied:
That makes sense, I'll work on it before we merge this!

vasiuadriana commented on content/module3/2-pre-requisites.md:
```diff
@@ -0,0 +1,60 @@
++++
```
I had the idea of including pre-requisites for the module. Basically anything that we need to have installed. What do you think @tmcalves ? 

I added the JDK as well, for completeness, even if that technically belongs to the first module. We can move it there if you prefer. The rest of the pre-requisites do belong here though.

tmcalves replied:
That's a really good idea, we never really mention local setup and I had issues on some pairings where people would use different environments so they'd be stuck.

vasiuadriana commented:
@tmcalves I created the structure of the module and the headers of everything I would like to cover. I think that with the 5 sections and 3 quite important topics - bash, java packaging + docker is a good size module. 
Happy to incorporate any suggestions you have.

I will try to keep each section concise and focus on getting things running, I won't go into additional details, because there is a lot of information anyway.

The flow of the module makes sense in my head, it aims to start from bash because we are using that in pre-requisites, than we use what we installed in the rest of the modules and at the end we have a running application locally -> which we can then use for the networking module :) Hopefully this makes sense.

tmcalves commented on content/module3/2-pre-requisites.md:
```diff
@@ -0,0 +1,60 @@
+Majority of the systems already come with a `JDK` installed. In order to check out the version you have installed, 
+open a terminal on your machine where you can use bash, and run `java --version`.
+
+In order to find instructions on how to install a version of `JDK`, you can have a look at the official documentation
```
Should we talk about java versions here? I'd probably mention what LTS versions are and which ones Java supports. I noticed the image uses JDK20 which is not LTS, should we make it so? Latest LTS is 21

vasiuadriana replied:
I added the LTS recommendation here and updated the IntelliJ screenshot to contain JDK21

tmcalves commented on content/module3/2-pre-requisites.md:
```diff
@@ -0,0 +1,60 @@
+
+## Build Tools
+
+In this module, we will build and package java applications using [Gradle](https://docs.gradle.org/current/userguide/getting_started_eng.html).
```
I did do a call on gradle and the current setup but never actually wrote anything so this is really useful

tmcalves commented on content/module3/2-pre-requisites.md:
```diff
@@ -0,0 +1,60 @@
+
+## Docker-compose
+
+We will cover the `docker-compose` basic usage when we cover [Running Containers with docker-compose](5-running-containers-docker-compose.md) but for now,
```
replace `docker-compose` with `docker compose`? since the former one is deprecated and is now included in the docker CLI

tmcalves commented on content/module3/3-package-java-app.md:
```diff
@@ -0,0 +1,28 @@
+In this module, we are going to use the `gradle` tool to create a java executable jar that contains all the dependencies.
+
+<TODO - add here link to the reference app with a basic build.gradle>
+<Question - shall we use java jar here?>
```
hm gradle is how you run locally, but production you'll always run the jar. Maybe mention both ways? If we're talking packaging I think it makes sense to show how to run it in production. Maybe also make the distinction that doing that way you don't need JDK, only need JRE, and what's the difference between both

vasiuadriana replied:
hey @tmcalves , I added some changes here https://github.com/Kalschatzi/learning_platform/pull/15/commits/6d469aec839ece4388a0b9f9a50184ad248a010b that make the distinction between `JDK` and `JRE` and improved the info related to development lifecycle - what you need when you build, versus running. If you are happy with this please close the comment.

tmcalves commented on content/module3/5-running-containers-docker-compose.md:
```diff
@@ -0,0 +1,11 @@
+weight = 5
++++
+
+## What is docker-compose?
```
same as previous comment, I'd replace with `docker compose` just to be the up to date syntax

tmcalves commented on content/module3/1-introduction-to-bash.md:
```diff
@@ -0,0 +1,11 @@
++++
+title = "Introduction to bash"
```
This is a highly request topic already 😄 going to be really useful

tmcalves reviewed (commented):
Answered questions and left small comments!
love the structure and pre-requisites are a great idea! need to add to other modules

vasiuadriana commented:
> Answered questions and left small comments! love the structure and pre-requisites are a great idea! need to add to other modules

Thanks Tiago! Questions and suggestions are great, I am working through them and also adding content for the rest of the sections. Whenever I am addressing the comments I mark them as resolved and after I have progressed more on the content and addressed all comments I will ask for a re-review. Exciting times :)

vasiuadriana commented:
hey @tmcalves - I decided to separate the bash introduction in it's own separate module because there are a few topics I want to cover there. 
This means that there is not much left on this module and is also more concise, which I like, because it allows us to incorporate any feedback we get. The things left are just the commands that need to be run - which depend on the reference app.

tmcalves commented on content/module3/1-pre-requisites.md:
```diff
@@ -0,0 +1,69 @@
+When you go through this module we recommend that you `gradle` both from command line and any IDE you might choose,
+so that you get familiar with the tool. 
+
+We recommend using the [Gradle Wrapper](https://docs.gradle.org/8.14/userguide/gradle_wrapper.html) in order to run gradle tasks.
```
I like we can now reference our own things eheh

tmcalves commented on content/module3/2-package-java-app.md:
```diff
@@ -0,0 +1,53 @@
+
+The pre-requisite for running a jar file is to have already a `JRE` (Java Runtime Environment) installed on the machine.
+
+What is the difference between `JDK` and `JRE`?
```
nice!

tmcalves commented on content/module3/3-containerization-with-docker.md:
```diff
@@ -0,0 +1,72 @@
+## Why using docker?
+
+Docker allows us to separate your application from the actual infrastructure. So you can run the same container anywhere: in the cloud, on a server somewhere or on your local machine.
+There are at least two major advantages that docker gives you when it comes to software development:
```
should we do any comparison with VMs? typical interview question, even on CECG, but not sure this is the right place yet

vasiuadriana replied:
I think it's a good idea to touch on this here, is the right place.

vasiuadriana replied:
I added a separate concise section to cover the main difference between them

tmcalves commented on content/module3/3-containerization-with-docker.md:
```diff
@@ -0,0 +1,72 @@
+- To run the image you created above, take the docker `IMAGE ID` displayed when you run `docker images` and do: `docker run <docker IMAGE ID>`. 
+- Now try and access the reference app status endpoint: http://localhost:8080/private/status. You will notice that you cannot connect to the server. 
+  That is because you haven't actually exposed a port on your localhost so that the docker container can be reached. 
+- To expose a port run `docker run -p 8080:8080 <docker IMAGE ID>`.
```
why do I need to expose the port? should we talk about docker network?

vasiuadriana replied:
I can mention the docker network here, I think is a good idea. I won't go into a lot of detail though because I think that would be too much for this module.

tmcalves replied:
yup, it's definitely a more complex topic, but maybe some light touch here and and link to more info can be useful

vasiuadriana replied:
I added an explanation and further links for this.

tmcalves reviewed (commented):
looking really good, great work!
left a couple of question

tmcalves reviewed (approved):
this looks great, awesome job 🎉
